### PR TITLE
Project Network Policy changes

### PR DESCRIPTION
**Pull request by leodotcloud**
https://github.com/rancher/rancher/issues/11145

### CODE REVIEW

File: pkg/controllers/user/networkpolicy/policy.go [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+	"strings"
+)
+
+const (
+	projectIDAnnotation   = "field.cattle.io/projectId"
+	projectIDNoColonLabel = "management.cattle.io/projectIdNoColon"
+)
+
+type mgr struct {
+	cluster   *config.UserContext
```
Comment (ibuildthecloud) [inline pkg/controllers/user/networkpolicy/policy.go]: avoid saving the whole context.  It's hard to determine your code dependencies when you have access to a global object like this.

File: pkg/controllers/user/networkpolicy/policy.go [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+
+	projectIDNoColon, ok := pnp.Labels[projectIDNoColonLabel]
+	if !ok {
+		return pnp, fmt.Errorf("couldn't find projectIDNoLabel for pnp=%+v", *pnp)
+	}
+	logrus.Debugf("pnplc Updated projectIDNoColon: %v", projectIDNoColon)
+
+	set := labels.Set(map[string]string{projectIDNoColonLabel: projectIDNoColon})
+
+	err := pnplc.mgr.cluster.K8sClient.NetworkingV1().NetworkPolicies("").DeleteCollection(nil, v1.ListOptions{LabelSelector: set.String()})
```
Comment (ibuildthecloud) [inline pkg/controllers/user/networkpolicy/policy.go]: You should be a good citizen and only delete what you create.  Just label your policies.

File: pkg/controllers/user/networkpolicy/policy.go [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+		return pnp, fmt.Errorf("couldn't find projectID for pnp=%+v", *pnp)
+	}
+	logrus.Debugf("pnplc Create projectIDNoColon: %v", projectIDNoColon)
+
+	err := programNetworkPolicy(pnplc.mgr, projectIDNoColon)
+	return pnp, err
+}
+
+func (pnplc *projectNetworkPolicyLifecycle) Updated(pnp *v3.ProjectNetworkPolicy) (*v3.ProjectNetworkPolicy, error) {
+	logrus.Infof("pnplc Updated pnp=%+v", pnp)
```
Comment (ibuildthecloud) [inline pkg/controllers/user/networkpolicy/policy.go]: This seems like copy and paste from Create

Comment (ibuildthecloud) [reply pkg/controllers/user/networkpolicy/policy.go]: Also note.  After create, Update is always called at least once.  So if you have no specific create logic, you can just leave Create empty

File: pkg/controllers/user/networkpolicy/policy.go [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+		return nil
+	}
+	return ps.mgr.createDefaultNetworkPolicy(p)
+}
+
+func (mgr *mgr) createDefaultNetworkPolicy(p *v3.Project) error {
+	if p == nil {
+		return nil
+	}
+	existingPolicies, err := mgr.pnpLister.List(p.Name, labels.NewSelector())
```
Comment (ibuildthecloud) [inline pkg/controllers/user/networkpolicy/policy.go]: There is a constant `labels.Everything`, you probably want to use that.

File: pkg/controllers/user/networkpolicy/policy.go [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+	if err != nil {
+		logrus.Errorf("error fetching existing project network policy: %v", err)
+		return nil
```
Comment (ibuildthecloud) [inline pkg/controllers/user/networkpolicy/policy.go]: why do you mask the error?

File: pkg/controllers/user/networkpolicy/policy.go [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+		pnpLister,
+		nsLister,
+	}
+
+	//plc := &projectLifecycle{mgr: mgr}
+	ps := &projectSyncer{mgr: mgr}
+	nslc := &namespaceLifecycle{mgr: mgr}
+	pnplc := &projectNetworkPolicyLifecycle{mgr: mgr}
+
+	cluster.Management.Management.Projects("").Controller().AddHandler("projectSyncer", ps.Sync)
```
Comment (ibuildthecloud) [inline pkg/controllers/user/networkpolicy/policy.go]: These should probably all be AddClusterScoped(Handler/Lifecycle)

File: pkg/controllers/user/networkpolicy/policy.go [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+	//cluster.Management.Management.Projects("").AddLifecycle("projectLifecycle", plc)
+	cluster.Management.Management.ProjectNetworkPolicies("").AddLifecycle("projectNetworkPolicyLifecycle", pnplc)
+	cluster.Core.Namespaces("").AddLifecycle("namespaceLifecycle", nslc)
+}
+
+func (pnplc *projectNetworkPolicyLifecycle) Create(pnp *v3.ProjectNetworkPolicy) (*v3.ProjectNetworkPolicy, error) {
+	logrus.Infof("pnplc Create pnp=%+v", *pnp)
+
+	projectIDNoColon, ok := pnp.Labels[projectIDNoColonLabel]
```
Comment (ibuildthecloud) [inline pkg/controllers/user/networkpolicy/policy.go]: pnp.Namespace is "projectIDNoColon"

File: pkg/controllers/user/networkpolicy/policy.go [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	networkingv1 "k8s.io/api/networking/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"strings"
+)
+
+const (
+	projectIDAnnotation   = "field.cattle.io/projectId"
+	projectIDNoColonLabel = "management.cattle.io/projectIdNoColon"
```
Comment (ibuildthecloud) [inline pkg/controllers/user/networkpolicy/policy.go]: this is not a good name.  What is this used for?

Comment (leodotcloud) [reply pkg/controllers/user/networkpolicy/policy.go]: I am adding a label to the network policy using the project id + namespace id but without colon which will be used later to do a lookup using the label.

Comment (ibuildthecloud) [reply pkg/controllers/user/networkpolicy/policy.go]: just change to `management.cattle.io/projectName`

File: pkg/controllers/user/networkpolicy/policy.go [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+		},
+	})
+	if err == nil {
+		logrus.Infof("Successfully created default network policy for project: %v", p.Name)
+	}
+	return err
+}
+
+func (nslc *namespaceLifecycle) Create(ns *corev1.Namespace) (*corev1.Namespace, error) {
+	logrus.Infof("nslc Create: %+v", *ns)
```
Comment (ibuildthecloud) [inline pkg/controllers/user/networkpolicy/policy.go]: Can you move the namespace labelers thingy to a completely separate controller, not in the networkpolicy package.

File: pkg/controllers/user/networkpolicy/policy.go [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+
+	err := pnplc.mgr.cluster.K8sClient.NetworkingV1().NetworkPolicies("").DeleteCollection(nil, v1.ListOptions{LabelSelector: set.String()})
+	if err != nil {
+		logrus.Errorf("error deleting network policies for projectIDNoColon=%v", projectIDNoColon)
+		return pnp, err
+	}
+	return pnp, nil
+}
+
+//func (plc *projectLifecycle) Create(p *v3.Project) (*v3.Project, error) {
```
Comment (ibuildthecloud) [inline pkg/controllers/user/networkpolicy/policy.go]: delete if not needed

**Comment by leodotcloud** [comment]
Closing in favor of https://github.com/rancher/rancher/pull/11615

**Pull request closed without merging**